### PR TITLE
allow staked nodes weight override

### PR DESCRIPTION
**Pull request by janlegner**
#### Problem
As per issue #26407: The validator may want to prioritize QUIC traffic from a known whitelist of sources and apply its own stake to those transactions.

#### Summary of Changes
- New parameter is added to the `solana-validator` program: `--staked_nodes_overrides <PATH/URL>`
- A configuration file is periodically fetched from the given location (filesystem or HTTP)
- The fetched configuration is applied to the staked nodes weight - effectively allowing whitelisting/prioritizing traffic from specified sources.

#### Config example
```yaml
auto_reload: true
stake_map:
  1.1.1.1: 1000000000000000
  2.2.2.2: 4000000000000000
```

### CODE REVIEW

**Comment by pgarg66** [comment]
It might be more useful to have overrides for pubkey -> stake mapping (instead of IP address -> stake). Most known connections will be from peer nodes that have active stake (e.g. previous leaders, or soon to be leaders).

**Comment by janlegner** [comment]
> It might be more useful to have overrides for pubkey -> stake mapping (instead of IP address -> stake). Most known connections will be from peer nodes that have active stake (e.g. previous leaders, or soon to be leaders). If the peer is staked, it's identity pubkey is being used to lookup its stake.

That is a good idea - but what about e.g. RPC pools without vote pubkey? :thinking: Maybe allowing both IP and pubkey would be a good idea?

**Comment by mvines** [comment]
I don't like the idea of a new service within the validator that tries to keep this list updated. Instead we only need a couple primitives:
1. New solana-validator command-line argument to provide the data file
2. A new admin ipc method to signal a reload of the data file (cc: https://github.com/solana-labs/solana/blob/ec36f0c5dffb36d639bc447a7552edbb8a241aa1/validator/src/admin_rpc_service.rs#L134)

**Comment by janlegner** [comment]
> I don't like the idea of a new service within the validator that tries to keep this list updated. Instead we only need a couple primitives:
> 
> 1. New solana-validator command-line argument to provide the data file
> 2. A new admin ipc method to signal a reload of the data file (cc: https://github.com/solana-labs/solana/blob/ec36f0c5dffb36d639bc447a7552edbb8a241aa1/validator/src/admin_rpc_service.rs#L134
>    )

Thanks for the suggestion, I did not think about using admin signals but that makes a lot of sense. Where do you think we should then move the logic loading the configuration from file/URL? Keeping it in `core/src/...` seems strange. I do not want to inflate `validator/src/main.rs` too much, so maybe `validator/src/lib.rs`?

**Comment by mvines** [comment]
> I do not want to inflate `validator/src/main.rs` too much, so maybe `validator/src/lib.rs`?

Yep that seems great. Putting it in `lib.rs` will also make it easier to add to `solana-test-validator` if somebody wants the functionality added there as well one day

**Comment by ochaloup** [comment]
I did an update behalf of @janlegner on this PR. The yaml configuration is loaded via RPC admin command or as a configuration option at the start of the validator program. The configuration works with the node identities now (/cc @pgarg66) instead of the IPs.

@mvines would you be so kind and provide me a feedback on this PR?

File: validator/src/admin_rpc_service.rs
Comment (mvines) [inline validator/src/admin_rpc_service.rs:464]: My preference is to remove the URL support entirely. Local files only. It's trivial for the caller to `wget` the file, or use whatever other standard url fetching tooling they prefer.

File: validator/src/main.rs
Comment (mvines) [inline validator/src/main.rs:1245]: It'd be nice to include some more info in the `.help()` about what impact this overrides may have for the validator operationally, either as an inline or a call out to external documentation.

File: core/src/staked_nodes_updater_service.rs
Comment (mvines) [inline core/src/staked_nodes_updater_service.rs:26]: I don't love how the placement of this struct here causes a number of extra downstream `use` of the `staked_nodes_updater_service` module, which really feels like it ought to be private to `solana-core`. Likewise, it necessitates a new `serde_yaml` dep for core.

Any way we can move the `serde_yaml` dep fully into `validator/` instead?

Also are these files going to be large? It's been my experience that `serde_yaml` is very slow on large files (relative to `serde_json`)

**Review by mvines — COMMENTED** [review]
nice progress!

File: validator/src/admin_rpc_service.rs
Comment (ochaloup) [reply validator/src/admin_rpc_service.rs]: +1, I changed for using only the path as you recommended.

File: validator/src/main.rs
Comment (ochaloup) [reply validator/src/main.rs]: I added the more instructive description. I will be glad if you can comment on the way I did.

I searched for the documentation page that the description could be elaborated (not sure if it's the way to go now) but I haven't found any CLI reference guide where I could add that to. The solana cli has got the reference page that seems to be generated from the code (https://docs.solana.com/cli/usage#solana-cli). The `solana-validator` has got no reference guide. Should that added to some specific guide with even more details?

File: core/src/staked_nodes_updater_service.rs
Comment (ochaloup) [reply core/src/staked_nodes_updater_service.rs]: The declaration was used to easily load various data from the struct. It was meant to be similar as other structures coming from `core` to other modules. As it's the `validator` module that needs to provide data to the `core service` I haven't thought an easy way to do so.

The struct could be defined without the `serde_yaml` dependency but then it still means multiple extra downstream `use`.
I haven't thought anything super clever and I eventually pass the `HashMap` as the shared structure for the data passing. I don't like the fact that any change of the format or adding types for overrides (e.g. the discussed IP) means a bit more changes. 
Is it fine to do it this way?

About the `serde_yaml`. We like the yaml as it's easy to work with and the configuration file is not expected to be large. So I left there the loading from yaml for now. If you consider it should really be the json, I can change it.

**Comment by ochaloup** [comment]
@mvines @pgarg66 As browsing through the code I would like to check with you one detail.

The goal of this PR is to override the stake amounts (as per pr#26407) to make a validator peer to be considered with some amount of the stake and then from that calculate QUIC streams and packet vote stage.
I did it now in way that the mapping is taken from the (https://github.com/solana-labs/solana/blob/v1.11.5/core/src/staked_nodes_updater_service.rs#L75) `tvu_peers()` and overrde is applied only if the validator `identity` is valid within the TVU ips. I think that this could be limiting. When the validator happens to be a valid TVU peer? For the QUIC stake overriding purpose of ip/id could I use rather the `cluster_info.all_peers()`?

**Comment by pgarg66** [comment]
> @mvines @pgarg66 As browsing through the code I would like to check with you one detail.
> 
> The goal of this PR is to override the stake amounts (as per pr#26407) to make a validator peer to be considered with some amount of the stake and then from that calculate QUIC streams and packet vote stage. I did it now in way that the mapping is taken from the (https://github.com/solana-labs/solana/blob/v1.11.5/core/src/staked_nodes_updater_service.rs#L75) `tvu_peers()` and overrde is applied only if the validator `identity` is valid within the TVU ips. I think that this could be limiting. When the validator happens to be a valid TVU peer? For the QUIC stake overriding purpose of ip/id could I use rather the `cluster_info.all_peers()`?
> 

It makes sense to use `all_peers()`. It'll allow a validator to allocate more QUIC streams to a node that's not a staked validator (e.g. an RPC node, or a client). 

> Btw. could be two validators running with the same IP? Could be then the `ip_to_stake` map in clash in values (https://github.com/solana-labs/solana/blob/v1.11.5/core/src/staked_nodes_updater_service.rs#L82)?

It's technically possible for two validators, e.g. behind NAT, using the same IP address. For this reason, we are not using ip_stake_map in QUIC. We have a mechanism to get peer's identity pubkey during QUIC connection setup process, and that's used for looking up id_to_stake (pubkey_stake_map) table.

**Comment by ochaloup** [comment]
> It makes sense to use `all_peers()`. It'll allow a validator to allocate more QUIC streams to a node that's not a staked validator (e.g. an RPC node, or a client).

 I see. Thank you @pgarg66 for great elaboration! I changed the code for `all_peers()` as suggested (if you can check it would be nice).

**Comment by pgarg66** [comment]
> > It makes sense to use `all_peers()`. It'll allow a validator to allocate more QUIC streams to a node that's not a staked validator (e.g. an RPC node, or a client).
> 
> 
> 
>  I see. Thank you @pgarg66 for this great elaboration! I changed the code for `all_peers()` as suggested (if you can check it would be nice).

Thanks for updating the PR. Looks good to me. Let's wait for @mvines to review it too, before we merge it.

**Review by mvines — APPROVED** [review]
lgtm, thanks!

**Pull request merged**
